network/src/dms2.rs

```diff
@@ -0,0 +1,670 @@
+use super::Storage;
+use super::*;
+use async_trait::async_trait;
+use eyre::eyre;
+use futures::future::join;
+use futures::prelude::*;
+use serde_tc::http::*;
+use serde_tc::{serde_tc_full, StubCall};
+use simperby_common::*;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::RwLock;
+
+const STATE_FILE_PATH: &str = "_state.json";
+type DmsKey = String;
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Message {
+    data: String,
+    height: BlockHeight,
+    dms_key: DmsKey,
+    signature: TypedSignature<(String, BlockHeight, DmsKey)>,
+}
+
+impl Message {
+    pub fn new(
+        data: String,
+        height: BlockHeight,
+        dms_key: DmsKey,
+        signature: TypedSignature<(String, BlockHeight, DmsKey)>,
+    ) -> Result<Self, CryptoError> {
+        signature.verify(&(data.clone(), height, dms_key.clone()))?;
+        Ok(Self {
+            data,
+            height,
+            dms_key,
+            signature,
+        })
+    }
+
+    pub fn data(&self) -> &str {
+        &self.data
+    }
+
+    pub fn signature(&self) -> &TypedSignature<(String, BlockHeight, DmsKey)> {
+        &self.signature
+    }
+}
+
+impl ToHash256 for Message {
+    fn to_hash256(&self) -> Hash256 {
+        Hash256::hash(serde_spb::to_vec(self).unwrap())
+    }
+}
+
+/// Decides whether a message should be accepted or not.
+pub trait MessageFilter: Send + Sync + 'static {
+    fn filter(&self, message: &Message) -> Result<(), String>;
+}
+
+/// A message before verification.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RawMessage {
+    pub data: String,
+    pub height: BlockHeight,
+    pub dms_key: DmsKey,
+    pub signature: TypedSignature<(String, BlockHeight, DmsKey)>,
+}
+
+impl RawMessage {
+    pub fn into_message(self) -> eyre::Result<Message> {
+        Message::new(self.data, self.height, self.dms_key, self.signature).map_err(|e| eyre!(e))
+    }
+
+    pub fn from_message(message: Message) -> Self {
+        RawMessage {
+            data: message.data().to_owned(),
+            height: message.height,
+            dms_key: message.dms_key.to_owned(),
+            signature: message.signature().to_owned(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct State {
+    pub dms_key: DmsKey,
+}
+
+pub struct DistributedMessageSet<S> {
```

Comment:
Why not trait bound here? `<S: Storage>`

Reply:
The trait bound is required by the implementation anyway. Thus, instantiating a `DistributedMessageSet` with the `S: !Storage` is allowed technically (which's what you were worried about, I guess) but that doesn't matter because there is no available method of it. I'm not sure about the practice on this case, so I'll just put one. (please check the common practice later)

Reply:
However AFAIK, it's desirable to have a trait bound only when it matters (i.e. for the implementation). If not, you have to repeat the trait bound everytime you face another trait impls. For example, we have to put the trait bound again on `impl<S> std::fmt::Debug` which doesn't actually require the `S` to be `Storage`.

Reply:
I thought we need to put the bound since this struct is designed to accept only `Storage`s, noticing the readers of that. However, I agree with you that it is desirable to specify the trait bound only when it matters, so let's leave this as it is.

Reply:
You can resolve this.